draco/analysis/flagging.py

```diff
@@ -1755,3 +1674,207 @@ def process(self, data):
+        Any weights with values less than this number times the average weight
+        will be set to zero. Default: 1e-6.
+    pols_to_flag : string, optional
+        Which polarizations to flag. "copol" only flags XX and YY baselines,
+        while "all" flags everything. Default: "all".
+    """
+
+    average_type = config.enum(["median", "mean"], default="median")
+    absolute_threshold = config.Property(proptype=float, default=1e-7)
+    relative_threshold = config.Property(proptype=float, default=1e-6)
+    pols_to_flag = config.enum(["all", "copol"], default="all")
+
+    def setup(self, telescope):
+        """Set the telescope model.
+
+        Parameters
+        ----------
+        telescope : TransitTelescope
+        """
+
+        self.telescope = io.get_telescope(telescope)
+
+    def process(
+        self,
+        stream: Union[
+            containers.TimeStream, containers.SiderealStream, andata.CorrData
+        ],
+    ) -> Union[containers.BaselineMask, containers.SiderealBaselineMask]:
+        """Construct mask and apply to input weights, if desired.
+
+        Parameters
+        ----------
+        stream : `.core.container` with `weight` attribute
+            Input container whose weights are used to construct the mask.
+
+        Returns
+        -------
+        out : `BaselineMask` or `SiderealBaselineMask`
+            The output baseline-dependent mask.
+        """
+        from mpi4py import MPI
+
+        # Only redistribute the weight dataset, because CorrData containers will have
+        # other parallel datasets without a stack axis
+        stream.weight.redistribute(axis=1)
+
+        # Make the output container, depending on input type
+        if "ra" in stream.axes:
+            mask_cont = containers.SiderealBaselineMask(
+                axes_from=stream, attrs_from=stream
+            )
+        elif "time" in stream.axes:
+            mask_cont = containers.BaselineMask(axes_from=stream, attrs_from=stream)
+        else:
+            raise TypeError(
+                f"Task requires TimeStream, SiderealStream, or CorrData. Got {type(stream)}"
+            )
+
+        # Redistribute output container along stack axis
+        mask_cont.redistribute("stack")
+
+        # Get local section of weights
+        local_weight = stream.weight[:].local_array
+
+        # Define averaging function that takes mean or median of nonzero elements.
+        # If all elements are zero, simply return zero.
+        average_func = np.median if self.average_type == "median" else np.mean
+        average_nonzero_func = (
+            lambda x: average_func(x[x != 0].astype(float))
+            if np.sum(x != 0) > 0
+            else 0.0
+        )
+
+        # Apply averaging function to weights for each baseline
+        average_weight = np.array(
+            [average_nonzero_func(w) for w in local_weight.transpose(1, 0, 2)],
+            dtype=local_weight.dtype,
+        )
+
```

Comment:
```suggestion
    average_weight = average_func(
        np.ma.array(local_weight, mask=(local_weight==0.0)), 
        axis=(0,2)
    ).data
```
This combined with using np.ma.mean and np.ma.median should do the same as the combo of the lambda function + average_weight loop above (np.ma.array shouldn't make a copy of the data).

Reply:
Thanks! I double-checked that this doesn't affect the results, so I've made this change